operator/pkg/clustermesh/clustermesh.go

```diff
@@ -0,0 +1,147 @@
+type clusterMesh struct {
+	// common implements the common logic to connect to remote clusters.
+	common common.ClusterMesh
+
+	Metrics Metrics
+	// globalServices is a list of all global services. The datastructure
+	// is protected by its own mutex inside the structure.
+	globalServices *globalServiceCache
+	stopCh         chan struct{}
+
+	storeFactory store.Factory
+
+	conurrentClusterMeshEndpointSync int
+
+	meshPodInformer     *meshPodInformer
+	meshNodeInformer    *meshNodeInformer
+	meshServiceInformer *meshServiceInformer
+
+	endpointSliceMeshController  *endpointslice.Controller
+	endpointSliceInformerFactory informers.SharedInformerFactory
+}
+
+func newClusterMesh(lc hive.Lifecycle, params clusterMeshParams) *clusterMesh {
+	if !params.Clientset.IsEnabled() || params.ClusterMeshConfig == "" {
+		return nil
+	}
+
+	cm := clusterMesh{
+		Metrics: params.Metrics,
+		globalServices: newGlobalServiceCache(
+			params.Metrics.TotalGlobalServices.WithLabelValues(params.ClusterInfo.Name),
+		),
+		stopCh:                           make(chan struct{}),
+		storeFactory:                     params.StoreFactory,
+		conurrentClusterMeshEndpointSync: params.Cfg.ConurrentClusterMeshEndpointSync,
+	}
+	cm.meshPodInformer = &meshPodInformer{GlobalServiceCache: cm.globalServices}
+	cm.meshServiceInformer = newMeshServiceInformer(cm.globalServices, params.Clientset)
+	cm.meshNodeInformer = newMeshNodeInformer()
+	cm.endpointSliceMeshController, cm.endpointSliceInformerFactory = newEndpointSliceMeshController(
+		params.Cfg,
+		cm.meshPodInformer,
+		cm.meshServiceInformer,
+		cm.meshNodeInformer,
+		params.Clientset,
+	)
+	cm.common = common.NewClusterMesh(common.Configuration{
+		Config:           params.Config,
+		ClusterInfo:      params.ClusterInfo,
+		NewRemoteCluster: cm.newRemoteCluster,
+		Metrics:          params.CommonMetrics,
+	})
+
+	lc.Append(&cm.common)
+	lc.Append(&cm)
+	return &cm
+}
+
+func (cm *clusterMesh) newRemoteCluster(name string, status common.StatusFunc) common.RemoteCluster {
+	rc := &remoteCluster{
+		name:         name,
+		mesh:         cm,
+		storeFactory: cm.storeFactory,
+		synced:       newSynced(),
+	}
+
+	rc.remoteServices = cm.storeFactory.NewWatchStore(
+		name,
+		func() store.Key { return new(serviceStore.ClusterService) },
+		&remoteServiceObserver{remoteCluster: rc},
+		store.RWSWithOnSyncCallback(func(ctx context.Context) { close(rc.synced.services) }),
+	)
+
+	return rc
+}
+
+func (cm *clusterMesh) Start(hive.HookContext) error {
+	// Run EndpointSlice sync
+	cm.endpointSliceInformerFactory.Start(cm.stopCh)
+	go cm.meshServiceInformer.Run(cm.stopCh)
+	cm.endpointSliceInformerFactory.WaitForCacheSync(cm.stopCh)
+
+	if !cache.WaitForCacheSync(cm.stopCh, cm.meshServiceInformer.HasSynced) {
+		return fmt.Errorf("WaitForCacheSync on service informer not successful")
+	}
+
+	go cm.endpointSliceMeshController.Run(context.Background(), cm.conurrentClusterMeshEndpointSync)
```

Comment:
Ohh yes very good catch thanks, at some point I was calling `(*clusterMesh).ServicesSynced` (https://github.com/MrFreezeex/cilium/blob/5222d983fdb716d65e7564cc2d6cd7ce9343ef33/operator/pkg/clustermesh/clustermesh.go#L125) but I somehow removed it for some reasons :thinking:. Do you think adding this back should be sufficient or there is something else that we would need to do?

AFAIK the datapath does call this as well :thinking:

Reply:
I think it depends on the intended behavior. I mean, should a misconfigured cluster block the synchronization of the endpoints from all other clusters? Ideally we would need something more granular to allow processing the updates from one cluster as soon as it is fully synchronized, but I'm honestly unsure if that is even possible with the upstream endpointslice controller. Maybe starting one controller per cluster, and keeping only a global GC logic to account for disconnections?

On a side note, any function waiting for cluster synchronization should not be called from a start hook, because that would block the processing of the subsequent start hooks.

Reply:
Oh ok yes I see what you mean. it should be possible in theory because the current code treats each cluster individually/triggers one reconciliation per cluster but I need to check how we can do an individual check to see if each cluster would be synced and how it would look like.

I have one concern doing this on an individual cluster though, it means that if a cluster is unreachable/misconfigured or whatever we probably need to add some form of timeout to still trigger a reconcilation of the endpoints somehow. Because if we don't do this the endpoints might never be removed...

Reply:
> it means that if a cluster is unreachable/misconfigured or whatever we probably need to add some form of timeout to still trigger a reconcilation of the endpoints somehow.

Yeah, that's another aspect to consider, and there could be different alternatives depending on whether we want to just preserve the epslices as they are (unless the cluster is completely removed), mark them as non-ready or remove them altogether after a certain timeout. I'm not even sure if there's a one size fit all solution in this case, as depending on the scenario one may prefer one solution over another. But I guess that for a first iteration whatever approach should work just fine.

Reply:
Thinking a bit more about this, the best idea is probably to keep the same behavior as the backends of global services (IIRC, we don't remove existing backends until all clusters are synchronized).

Reply:
This should be addressed but I am not marking it as resolved for visibility as it is somewhat important